sql/core/src/main/scala/org/apache/spark/sql/execution/joins/ShuffledJoin.scala

```diff
@@ -40,4 +43,43 @@ trait ShuffledJoin extends BaseJoinExec {
       throw new IllegalArgumentException(
         s"ShuffledJoin should not take $x as the JoinType")
   }
+
+  /**
+   * Creates variables and declarations for attributes in row.
+   *
+   * In order to defer the access after condition and also only access once in the loop,
+   * the variables should be declared separately from accessing the columns, we can't use the
+   * codegen of BoundReference here.
+   */
+  protected def createVars(
```

Comment:
Original `createLeftVars` is created to defer accessing of row fields after condition evaluation. But I look at the usage of this `createVars` in `HashJoin`, I don't see such thing. If you don't do defer there, you can simply use `BoundReference` codegen, it is much simpler.

Reply:
@viirya - I was actually originally using `BoundReference`, but got compilation error with variable redefinition. E.g.,  for code branch `c21:codegen-fail` ([with change to `BoundReference` compare to this PR]( https://github.com/c21/spark/compare/codegen...codegen-fail)), and [example query in `JoinBenchmark`](https://github.com/apache/spark/blob/master/sql/core/src/test/scala/org/apache/spark/sql/execution/benchmark/JoinBenchmark.scala#L153-L167).

The `BoundReference` version generated code like this:

```
while (shj_streamedInput_0.hasNext()) {
  shj_streamedRow_0 = (InternalRow) shj_streamedInput_0.next();
  
  // generate join key for stream side
  long shj_value_0 = shj_streamedRow_0.getLong(0); // 1st definition here
  
  // find matches from HashRelation
  scala.collection.Iterator shj_matches_0 = false ? null : (scala.collection.Iterator)shj_relation_0.get(shj_value_0);
  while (shj_matches_0.hasNext()) {
    UnsafeRow shj_matched_0 = (UnsafeRow) shj_matches_0.next();
    long shj_value_0 = shj_streamedRow_0.getLong(0); // 2nd definition here and compilation error
    shj_mutableStateArray_0[0].write(0, shj_value_0);           
  }
}
```

So basically the variable `shj_value_0` here (stream side key ) needs to be defer accessed in [`HashJoin.consume()`](https://github.com/c21/spark/blob/codegen/sql/core/src/main/scala/org/apache/spark/sql/execution/joins/HashJoin.scala#L441), and it was first accessed in [`HashJoin.genStreamSideJoinKey()`](https://github.com/c21/spark/blob/codegen/sql/core/src/main/scala/org/apache/spark/sql/execution/joins/HashJoin.scala#L326). So it seems that `BoundReference` not work for me out of box.

Let me know if it makes sense or there would be any other better approach, thanks.